src/client-side-encryption/autoEncrypter.ts

```diff
@@ -0,0 +1,641 @@
+    }
+  }
+
+  /**
+   * Initializes the auto encrypter by spawning a mongocryptd and connecting to it.
+   *
+   * This function is a no-op when bypassSpawn is set or the crypt shared library is used.
+   */
+  init(callback: Callback<MongoClient>) {
+    if (this._bypassMongocryptdAndCryptShared || this.cryptSharedLibVersionInfo) {
+      return callback();
+    }
+    if (!this._mongocryptdManager) {
+      return callback(
+        new MongoRuntimeError(
+          'Reached impossible state: mongocryptdManager is undefined when neither bypassSpawn nor the shared lib are specified.'
+        )
+      );
+    }
+    if (!this._mongocryptdClient) {
+      return callback(
+        new MongoRuntimeError(
+          'Reached impossible state: mongocryptdClient is undefined when neither bypassSpawn nor the shared lib are specified.'
+        )
+      );
+    }
+    const _callback = (err?: AnyError, res?: MongoClient) => {
+      if (
+        err &&
+        err.message &&
+        (err.message.match(/timed out after/) || err.message.match(/ENOTFOUND/))
+      ) {
+        callback(
+          new MongoError(
+            'Unable to connect to `mongocryptd`, please make sure it is running or in your PATH for auto-spawn'
+          )
+        );
+        return;
+      }
+
+      callback(err, res);
+    };
+
+    if (this._mongocryptdManager.bypassSpawn) {
+      this._mongocryptdClient.connect().then(
+        result => {
+          return _callback(undefined, result);
+        },
+        error => {
+          _callback(error, undefined);
+        }
+      );
+      return;
+    }
+
+    this._mongocryptdManager.spawn(() => {
+      if (!this._mongocryptdClient) {
+        return callback(
+          new MongoRuntimeError(
+            'Reached impossible state: mongocryptdClient is undefined after spawning libmongocrypt.'
+          )
+        );
+      }
+      this._mongocryptdClient.connect().then(
+        result => {
+          return _callback(undefined, result);
+        },
+        error => {
+          _callback(error, undefined);
+        }
+      );
+    });
+  }
+
+  /**
+   * Cleans up the `_mongocryptdClient`, if present.
+   */
+  teardown(force: boolean, callback: Callback<void>) {
+    if (this._mongocryptdClient) {
+      this._mongocryptdClient.close(force).then(
+        result => {
+          return callback(undefined, result);
+        },
+        error => {
+          callback(error);
+        }
+      );
+    } else {
+      callback();
+    }
+  }
+
+  encrypt(ns: string, cmd: Document, callback: Callback<Document | Uint8Array>): void;
+  encrypt(
+    ns: string,
+    cmd: Document,
+    options: CommandOptions,
+    callback: Callback<Document | Uint8Array>
+  ): void;
+  /**
+   * Encrypt a command for a given namespace.
+   */
+  encrypt(
+    ns: string,
+    cmd: Document,
+    options?: CommandOptions | Callback<Document | Uint8Array>,
```

Comment:
`encrypt` is only called in one location (CryptoConnection.command) and we pass the command arguments in.